lib/TPP/ToBlockLayoutAndBack.cpp

```diff
@@ -557,6 +557,62 @@ struct PropagateThroughPadOp : public OpRewritePattern<tensor::PadOp> {
   }
 };
 
+//===----------------------------------------------------------------------===//
+// BubbleUpThroughFillOp
+//===----------------------------------------------------------------------===//
+
+// Attempt to avoid packing a fill op. Instead create a 'packed' fill.
+// %0 = tensor.empty
+// %packed = tensor.empty
+// %1 = linalg.fill ins(%cst) outs(%0)
+// %2 = tensor.pack %1 into %packed
+// %3 = some_packed_op %2
+// %4 = tensor.unpack %3 into %1
+// --->
+// %0 = tensor.empty
+// %1 = linalg.fill ins(%cst) outs (%packed)
+// %2 = some_packed_op %1
+// %3 = tensor.unpack %2 into %0
+//
+struct BubbleUpThroughFillOp : public OpRewritePattern<tensor::PackOp> {
+  using OpRewritePattern<tensor::PackOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::PackOp packOp,
+                                PatternRewriter &rewriter) const override {
+    Value source = packOp.getSource();
+    auto fillOp = source.getDefiningOp<linalg::FillOp>();
+    if (!fillOp)
+      return failure();
+
+    Value fillRes = fillOp.getResult(0);
+    auto users = fillRes.getUsers();
+
+    // Only two users: a pack and an unpack.
+    if (std::distance(users.begin(), users.end()) != 2)
```

Comment:
I don't think presence of `unpack` (or another user) should be strictly required.

Reply:
Yes, it does. The fillOp must be dead otherwise you introduce re-computation.

Reply:
Ah I see, I initially thought the idea was to simply fold a `pack` into producing `fill` if there are no other users. But it is a bit more specific pattern.

Reply:
yeah, fairly specific.